gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/Traversal.java

```diff
@@ -254,9 +254,14 @@ public default void forEachRemaining(final Consumer<? super E> action) {
         }
     }
 
+    /**
+     * Releases resources opened in any steps that implement {@link AutoCloseable}.
+     */
     @Override
     public default void close() throws Exception {
```

Comment:
This is bad. `Steps` that are `TraversalParents` are responsible for propagating operations on them to those below. I would do the following:

```
for(final Step<?,?> step : this.getSteps()) {
  if(step instanceof AutoCloseable)
    step.close();
}
```

Next, I would have `TraversalParent`  extend `AutoCloseable` with its default `close()` method being:

```
for(final Traversal.Admin<?,?> traversal : this.getLocalChildren()) {
  traversal.close()
}

for(final Traversal.Admin<?,? traversal: this.getGlobalChildren()) {
  traversal.close();
}
```